streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -444,6 +444,18 @@ private void handleStreamsUncaughtException(final Throwable throwable,
                     "The old handler will be ignored as long as a new handler is set.");
         }
         switch (action) {
+            case REPLACE_THREAD:
+                StreamThread deadThread = (StreamThread) threads.stream().filter(n -> n.getName().equals(Thread.currentThread().getName())).toArray()[0];
+                threads.remove(deadThread);
```

Comment:
removing from the thread list does 2 things

1. keeps DEAD threads out of the list as kip-663 dictates
2. ensures the next thread has the same name

Reply:
To ensure #2 holds, do we need to have a mutex on thread creation? Or does it not matter?

Reply:
It won't matter. Ensures is too strong of a word. IF there has been other threads removed before this it may have a different name. However this ensures that the replacement thread will never have a thread index larger than the number of threads